File: cpplib_old_eroberts/cslib/tokenscanner.c
/*
 * File: tokenscanner.c
 * --------------------
 * Implementation for the TokenScanner class.
 */

#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include "cslib.h"
#include "strbuf.h"
#include "strlib.h"
#include "tokenscanner.h"
#include "unittest.h"

/*
 * Type: StringCell
 * ----------------
 * This type is used to construct linked lists of cells, which are used
 * to represent both the stack of saved tokens and the set of defined
 * operators.
 */

typedef struct StringCell {
   string str;
   struct StringCell *link;
} StringCell;

/*
 * Type: NumberScannerState
 * ------------------------
 * This type is used to enumerate the finite state machine used to read
 * numbers.
 */

typedef enum {
   INITIAL_STATE,
   BEFORE_DECIMAL_POINT,
   AFTER_DECIMAL_POINT,
   STARTING_EXPONENT,
   FOUND_EXPONENT_SIGN,
   SCANNING_EXPONENT,
   FINAL_STATE
} NumberScannerState;

/*
 * Type: TokenScannerCDT
 * ---------------------
 * This type represents the concrete structure of a scanner.
 */

struct TokenScannerCDT {
   string buffer;                   /* Source for string input      */
   int len;                         /* Length of buffer             */
   int cp;                          /* Current buffer position      */
   FILE *infile;                    /* Source for file input        */
   bool charIsSaved;                /* Char has been saved to file  */
   bool ignoreWhitespaceFlag;       /* Scanner ignores whitespace   */
   bool ignoreCommentsFlag;         /* Scanner ignores comments     */
   bool scanNumbersFlag;            /* Scanner parses numbers       */
   bool scanStringsFlag;            /* Scanner parses strings       */
   StringCell *savedTokens;         /* Stack of saved tokens        */
   StringCell *operators;           /* List of multichar operators  */
   StringBuffer wordChars;          /* Additional word characters   */
   StringBuffer savedChars;         /* Stack of saved characters    */
   StringBuffer sb;                 /* Buffer for creating strings  */
};

/* Private method prototypes */

static void initScanner(TokenScanner scanner);
static void skipSpaces(TokenScanner scanner);
static string scanWord(TokenScanner scanner);
static string scanNumber(TokenScanner scanner);
static string scanString(TokenScanner scanner);
static bool isOperator(TokenScanner scanner, string op);
static bool isOperatorPrefix(TokenScanner scanner, string op);
static void freeStringList(StringCell *cp);

/* Exported entries */

TokenScanner newTokenScanner(void) {
   TokenScanner scanner;

   scanner = newBlock(TokenScanner);
   initScanner(scanner);
   setInputString(scanner, "");
   return scanner;
}

void freeTokenScanner(TokenScanner scanner) {
   freeStringList(scanner->savedTokens);
   freeStringList(scanner->operators);
   freeStringBuffer(scanner->sb);
   freeStringBuffer(scanner->savedChars);
   freeStringBuffer(scanner->wordChars);
   freeBlock(scanner);
}

void setInputString(TokenScanner scanner, string str) {
   scanner->buffer = str;
   scanner->len = stringLength(str);
   scanner->cp = 0;
   scanner->infile = NULL;
   freeStringList(scanner->savedTokens);
   scanner->savedTokens = NULL;
}

void setInputFile(TokenScanner scanner, FILE *infile) {
   scanner->infile = infile;
   scanner->charIsSaved = false;
   freeStringList(scanner->savedTokens);
   scanner->savedTokens = NULL;
}

bool hasMoreTokens(TokenScanner scanner) {
   string token;

   token = nextToken(scanner);
   saveToken(scanner, token);
   return stringLength(token) != 0;
}

string nextToken(TokenScanner scanner) {
   StringCell *cp;
   string token;
   int ch, prev;

   if (scanner->savedTokens != NULL) {
      cp = scanner->savedTokens;
      token = cp->str;
      scanner->savedTokens = cp->link;
      freeBlock(cp);
      return token;
   }
   while (true) {
      if (scanner->ignoreWhitespaceFlag) skipSpaces(scanner);
      ch = getChar(scanner);
      if (ch == '/' && scanner->ignoreCommentsFlag) {
         ch = getChar(scanner);
         if (ch == '/') {
            while (true) {
               ch = getChar(scanner);
               if (ch == '\n' || ch == '\r' || ch == EOF) break;
            }
            continue;
         } else if (ch == '*') {
            prev = EOF;
            while (true) {
               ch = getChar(scanner);
               if (ch == EOF || (prev == '*' && ch == '/')) break;
               prev = ch;
            }
            continue;
         }
         ungetChar(scanner, ch);
         ch = '/';
      }
      if (ch == EOF) return copyString("");
      if ((ch == '"' || ch == '\'') && scanner->scanStringsFlag) {
         ungetChar(scanner, ch);
         return scanString(scanner);
      }
      if (isdigit(ch) && scanner->scanNumbersFlag) {
         ungetChar(scanner, ch);
         return scanNumber(scanner);
      }
      if (isWordCharacter(scanner, ch)) {
         ungetChar(scanner, ch);
         return scanWord(scanner);
      }
      clear(scanner->sb);
      pushChar(scanner->sb, ch);
      while (isOperatorPrefix(scanner, getString(scanner->sb))) {
         ch = getChar(scanner);
         if (ch == EOF) break;
         pushChar(scanner->sb, ch);
      }
      while (size(scanner->sb) > 1
             && !isOperator(scanner, getString(scanner->sb))) {
         ungetChar(scanner, ch);
         popChar(scanner->sb);
      }
      return copyString(getString(scanner->sb));
   }
}

void saveToken(TokenScanner scanner, string token) {
   StringCell *cp;

   cp = newBlock(StringCell *);
   cp->str = token;
   cp->link = scanner->savedTokens;
   scanner->savedTokens = cp;
}

void ignoreWhitespace(TokenScanner scanner) {
   scanner->ignoreWhitespaceFlag = true;
}

void ignoreComments(TokenScanner scanner) {
   scanner->ignoreCommentsFlag = true;
}

void scanNumbers(TokenScanner scanner) {
   scanner->scanNumbersFlag = true;
}

void scanStrings(TokenScanner scanner) {
   scanner->scanStringsFlag = true;
}

void addWordCharacters(TokenScanner scanner, string str) {
   appendString(scanner->wordChars, str);
}

void addOperator(TokenScanner scanner, string op) {
   StringCell *cp;

   cp = newBlock(StringCell *);
   cp->str = op;
   cp->link = scanner->operators;
   scanner->operators = cp;
}

int getPosition(TokenScanner scanner) {
   int pos;

   if (scanner->infile == NULL) {
      pos = scanner->cp;
   } else {
      pos = (int) ftell(scanner->infile);
   }
   if (scanner->savedTokens != NULL) {
      if (scanner->savedTokens->link != NULL) return -1;
      pos -= stringLength(scanner->savedTokens->str);
   }
   return pos;
}

bool isWordCharacter(TokenScanner scanner, char ch) {
   return isalnum(ch) || findChar(ch, getString(scanner->wordChars), 0) != -1;
}

void verifyToken(TokenScanner scanner, string expected) {
   string token;

   token = nextToken(scanner);
   if (!stringEqual(token, expected)) {
      if (stringEqual(token, "\n")) token = "end of line";
      if (stringEqual(expected, "\n")) expected = "end of line";
      error("Found \"%s\" when expecting \"%s\"", token, expected);
   }
   freeBlock(token);
}

TokenType getTokenType(TokenScanner scanner, string token) {
   char ch;

   if (stringLength(token) == 0) return (TokenType) EOF;
   ch = token[0];
   if (isspace(ch)) return SEPARATOR;
   if ((ch == '"' || ch == '\'') && stringLength(token) > 1) return STRING;
   if (isdigit(ch)) return NUMBER;
   if (isWordCharacter(scanner, ch)) return WORD;
   return OPERATOR;
}


int getChar(TokenScanner scanner) {
   if (!isEmpty(scanner->savedChars)) return popChar(scanner->savedChars);
   if (scanner->infile == NULL) {
      if (scanner->cp >= scanner->len) {
         return EOF;
      } else {
         return scanner->buffer[scanner->cp++];
      }
   } else {
      scanner->charIsSaved = false;
      return getc(scanner->infile);
   }
}

void ungetChar(TokenScanner scanner, int ch) {
   if (scanner->infile == NULL) {
      if (scanner->cp == 0) error("Internal error: pushback before start");
      scanner->cp--;
   } else {
      if (scanner->charIsSaved) {
         pushChar(scanner->savedChars, (char) ch);
      } else {
         ungetc(ch, scanner->infile);
         scanner->charIsSaved = true;
      }
   }
}

string getStringValue(string token) {
   StringBuffer sb;
   int i, base, start, finish, result, digit;
   string str;
   char ch;

   sb = newStringBuffer();
   start = 0;
   finish = stringLength(token);
   if (finish > 1 && (token[0] == '"' || token[0] == '\'')) {
      start = 1;
      finish--;
   }
   for (i = start; i < finish; i++) {
      ch = token[i];
      if (ch == '\\') {
         ch = token[++i];
         if (isdigit(ch) || ch == 'x') {
            base = 8;
            if (ch == 'x') {
               base = 16;
               i++;
            }
            result = 0;
            digit = 0;
            while (i < finish) {
               ch = token[i];
               if (isdigit(ch)) {
                  digit = ch - '0';
               } else if (isalpha(ch)) {
                  digit = toupper(ch) - 'A' + 10;
               } else {
                  digit = base;
               }
               if (digit >= base) break;
               result = base * result + digit;
               i++;
            }
            ch = (char) result;
            i--;
         } else {
            switch (ch) {
             case 'a': ch = '\a'; break;
             case 'b': ch = '\b'; break;
             case 'f': ch = '\f'; break;
             case 'n': ch = '\n'; break;
             case 'r': ch = '\r'; break;
             case 't': ch = '\t'; break;
             case 'v': ch = '\v'; break;
             case '"': ch = '"'; break;
             case '\'': ch = '\''; break;
             case '\\': ch = '\\'; break;
            }
         }
      }
      pushChar(sb, ch);
   }
   str = copyString(getString(sb));
   freeStringBuffer(sb);
   return str;
}

/* Private methods */

static void initScanner(TokenScanner scanner) {
   scanner->ignoreWhitespaceFlag = false;
   scanner->ignoreCommentsFlag = false;
   scanner->scanNumbersFlag = false;
   scanner->scanStringsFlag = false;
   scanner->operators = NULL;
   scanner->savedTokens = NULL;
   scanner->sb = newStringBuffer();
   scanner->savedChars = newStringBuffer();
   scanner->wordChars = newStringBuffer();
}

/*
 * Implementation notes: skipSpaces
 * --------------------------------
 * Advances the position of the scanner until the current character is
 * not a whitespace character.  The string str is used to check that
 * a whitespace character is not included explicitly as an operator.
 */

static void skipSpaces(TokenScanner scanner) {
   char str[2];
   int ch;

   while (true) {
      ch = getChar(scanner);
      if (ch == EOF) return;
      str[0] = ch;
      str[1] = '\0';
      if (!(isspace(ch) && !isOperator(scanner, str))) {
         ungetChar(scanner, ch);
         return;
      }
   }
}

/*
 * Implementation notes: scanWord
 * ------------------------------
 * Reads characters until the scanner reaches the end of a sequence
 * of word characters.
 */

static string scanWord(TokenScanner scanner) {
   int ch;

   clear(scanner->sb);
   while (true) {
      ch = getChar(scanner);
      if (ch == EOF) break;
      if (!isWordCharacter(scanner, ch)) {
         ungetChar(scanner, ch);
         break;
      }
      pushChar(scanner->sb, (char) ch);
   }
   return copyString(getString(scanner->sb));
}

/*
 * Implementation notes: scanNumber
 * --------------------------------
 * Reads characters until the scanner reaches the end of a legal number.
 * The function operates by simulating what computer scientists
 * call a finite-state machine.  The program uses the variable
 * <code>state</code> to record the history of the process and
 * determine what characters would be legal at this point in time.
 */

static string scanNumber(TokenScanner scanner) {
   NumberScannerState state;
   int ch, xch;

   clear(scanner->sb);
   state = INITIAL_STATE;
   while (state != FINAL_STATE) {
      ch = getChar(scanner);
      xch = 'e';
      switch (state) {
       case INITIAL_STATE:
         if (!isdigit(ch)) {
            error("Internal error: illegal call to scanNumber");
         }
         state = BEFORE_DECIMAL_POINT;
         break;
       case BEFORE_DECIMAL_POINT:
         if (ch == '.') {
            state = AFTER_DECIMAL_POINT;
         } else if (ch == 'E' || ch == 'e') {
            state = STARTING_EXPONENT;
            xch = ch;
         } else if (!isdigit(ch)) {
            if (ch != EOF) ungetChar(scanner, ch);
            state = FINAL_STATE;
         }
         break;
       case AFTER_DECIMAL_POINT:
         if (ch == 'E' || ch == 'e') {
            state = STARTING_EXPONENT;
            xch = ch;
         } else if (!isdigit(ch)) {
            if (ch != EOF) ungetChar(scanner, ch);
            state = FINAL_STATE;
         }
         break;
       case STARTING_EXPONENT:
         if (ch == '+' || ch == '-') {
            state = FOUND_EXPONENT_SIGN;
         } else if (isdigit(ch)) {
            state = SCANNING_EXPONENT;
         } else {
            if (ch != EOF) ungetChar(scanner, ch);
            ungetChar(scanner, ch);
            state = FINAL_STATE;
         }
         break;
       case FOUND_EXPONENT_SIGN:
         if (isdigit(ch)) {
            state = SCANNING_EXPONENT;
         } else {
            if (ch != EOF) ungetChar(scanner, ch);
            ungetChar(scanner, ch);
            ungetChar(scanner, ch);
            state = FINAL_STATE;
         }
         break;
       case SCANNING_EXPONENT:
         if (!isdigit(ch)) {
            if (ch != EOF) ungetChar(scanner, ch);
            state = FINAL_STATE;
         }
         break;
       default:
         state = FINAL_STATE;
         break;
      }
      if (state != FINAL_STATE) {
         pushChar(scanner->sb, (char) ch);
      }
   }
   return copyString(getString(scanner->sb));
}

/*
 * Implementation notes: scanString
 * --------------------------------
 * Reads and returns a quoted string from the scanner, continuing until
 * it scans the matching delimiter.  The scanner generates an error if
 * there is no closing quotation mark before the end of the input.
 */

static string scanString(TokenScanner scanner) {
   bool escape;
   char delim;
   int ch;

   clear(scanner->sb);
   delim = getChar(scanner);
   pushChar(scanner->sb, delim);
   escape = false;
   while (true) {
      ch = getChar(scanner);
      if (ch == EOF) error("TokenScanner found unterminated string");
      if (ch == delim && !escape) break;
      escape = (ch == '\\') && !escape;
      pushChar(scanner->sb, (char) ch);
   }
   pushChar(scanner->sb, delim);
   return copyString(getString(scanner->sb));
}

/*
 * Implementation notes: isOperator, isOperatorPrefix
 * --------------------------------------------------
 * These methods search the list of operators and return true if the
 * specified operator is either in the list or a prefix of an operator
 * in the list, respectively.  This code could be made considerably more
 * efficient by implementing operators as a trie.
 */

static bool isOperator(TokenScanner scanner, string op) {
   StringCell *cp;

   for (cp = scanner->operators; cp != NULL; cp = cp->link) {
      if (stringEqual(op, cp->str)) return true;
   }
   return false;
}

static bool isOperatorPrefix(TokenScanner scanner, string op) {
   StringCell *cp;

   for (cp = scanner->operators; cp != NULL; cp = cp->link) {
      if (startsWith(cp->str, op)) return true;
   }
   return false;
}

static void freeStringList(StringCell *cp) {
   StringCell *next;
   while (cp != NULL) {
      next = cp->link;
      freeBlock(cp);
      cp = next;
   }
}

/**********************************************************************/
/* Unit test for the tokenscanner module                              */
/**********************************************************************/

#ifndef _NOTEST_

/* Unit test */

static void testStringScanner(void);
static void testStreamScanner(void);
static void testScannerNoOptions(void);
static void testScannerLanguageOptions(void);
static void testScanNumbers(void);
static void testScanEscapeSequences(void);

void testTokenScannerModule(void) {
   testStringScanner();
   testStreamScanner();
   testScanEscapeSequences();
}

static void testStringScanner(void) {
   testScannerNoOptions();
   testScannerLanguageOptions();
   testScanNumbers();
}

static void testStreamScanner(void) {
   TokenScanner scanner;
   FILE *infile;

   trace(scanner = newTokenScanner());
   trace(infile = fopen("tokenscanner.h", "r"));
   test(infile == NULL, false);
   if (infile == NULL) return;
   trace(setInputFile(scanner, infile));
   test(hasMoreTokens(scanner), true);
   test(nextToken(scanner), "/");
   test(nextToken(scanner), "*");
   test(hasMoreTokens(scanner), true);
   test(nextToken(scanner), "\n");
   test(hasMoreTokens(scanner), true);
   test(nextToken(scanner), " ");
   test(nextToken(scanner), "*");
   test(nextToken(scanner), " ");
   test(nextToken(scanner), "File");
   trace(ignoreWhitespace(scanner));
   trace(scanStrings(scanner));
   test(nextToken(scanner), ":");
   test(nextToken(scanner), "tokenscanner");
   test(nextToken(scanner), ".");
   test(nextToken(scanner), "h");
   test(nextToken(scanner), "*");
   test(nextToken(scanner), "-");
   trace(fclose(infile));
}

static void testScannerNoOptions(void) {
   TokenScanner scanner;
   string token;

   trace(scanner = newTokenScanner());
   trace(setInputString(scanner, "cout << \"pi = \" << 3.14159265 << endl;"));
   test(hasMoreTokens(scanner), true);
   test(token = nextToken(scanner), "cout");
   test(getTokenType(scanner, token), (int) WORD);
   test(token = nextToken(scanner), " ");
   test(getTokenType(scanner, token), (int) SEPARATOR);
   test(token = nextToken(scanner), "<");
   test(getTokenType(scanner, token), (int) OPERATOR);
   test(nextToken(scanner), "<");
   test(nextToken(scanner), " ");
   test(nextToken(scanner), "\"");
   test(nextToken(scanner), "pi");
   test(nextToken(scanner), " ");
   test(nextToken(scanner), "=");
   test(nextToken(scanner), " ");
   test(nextToken(scanner), "\"");
   test(nextToken(scanner), " ");
   test(nextToken(scanner), "<");
   test(nextToken(scanner), "<");
   test(nextToken(scanner), " ");
   test(token = nextToken(scanner), "3");
   test(getTokenType(scanner, token), (int) NUMBER);
   test(nextToken(scanner), ".");
   test(nextToken(scanner), "14159265");
   test(nextToken(scanner), " ");
   test(nextToken(scanner), "<");
   test(nextToken(scanner), "<");
   test(nextToken(scanner), " ");
   test(nextToken(scanner), "endl");
   test(nextToken(scanner), ";");
   test(hasMoreTokens(scanner), false);
   test(token = nextToken(scanner), "");
   test(getTokenType(scanner, token), (int) EOF);
}

static void testScannerLanguageOptions(void) {
   TokenScanner scanner;
   string token;

   trace(scanner = newTokenScanner());
   trace(ignoreWhitespace(scanner));
   trace(scanStrings(scanner));
   trace(scanNumbers(scanner));
   trace(addWordCharacters(scanner, "_"));
   trace(addOperator(scanner, "=="));
   trace(addOperator(scanner, "!="));
   trace(addOperator(scanner, "<="));
   trace(addOperator(scanner, ">="));
   trace(addOperator(scanner, "<<"));
   trace(addOperator(scanner, ">>"));
   trace(addOperator(scanner, "++"));
   trace(addOperator(scanner, "--"));
   trace(addOperator(scanner, "->"));
   trace(addOperator(scanner, "&&"));
   trace(addOperator(scanner, "||"));
   trace(setInputString(scanner, "cout << \"pi = \" << 3.14159265 << endl;"));
   test(hasMoreTokens(scanner), true);
   test(nextToken(scanner), "cout");
   test(nextToken(scanner), "<<");
   test(token = nextToken(scanner), "\"pi = \"");
   test(getTokenType(scanner, token), (int) STRING);
   test(nextToken(scanner), "<<");
   test(token = nextToken(scanner), "3.14159265");
   test(getTokenType(scanner, token), (int) NUMBER);
   test(nextToken(scanner), "<<");
   test(nextToken(scanner), "endl");
   test(nextToken(scanner), ";");
   test(hasMoreTokens(scanner), false);
   test(nextToken(scanner), "");
   trace(setInputString(scanner, "if (a_1<=b+++_c &&&x!=4)"));
   test(nextToken(scanner), "if");
   test(nextToken(scanner), "(");
   test(token = nextToken(scanner), "a_1");
   test(getTokenType(scanner, token), (int) WORD);
   test(nextToken(scanner), "<=");
   test(nextToken(scanner), "b");
   test(nextToken(scanner), "++");
   test(nextToken(scanner), "+");
   test(token = nextToken(scanner), "_c");
   test(getTokenType(scanner, token), (int) WORD);
   test(nextToken(scanner), "&&");
   test(nextToken(scanner), "&");
   test(nextToken(scanner), "x");
   test(nextToken(scanner), "!=");
   test(nextToken(scanner), "4");
   test(nextToken(scanner), ")");
   test(hasMoreTokens(scanner), false);
   test(nextToken(scanner), "");
}

static void testScanNumbers(void) {
   TokenScanner scanner;

   trace(scanner = newTokenScanner());
   trace(ignoreWhitespace(scanner));
   trace(scanNumbers(scanner));
   trace(setInputString(scanner, "42 73x 3.14159265 6.28E-25"));
   test(getChar(scanner), '4');
   trace(ungetChar(scanner, '4'));
   test(nextToken(scanner), "42");
   test(nextToken(scanner), "73");
   test(nextToken(scanner), "x");
   test(nextToken(scanner), "3.14159265");
   test(nextToken(scanner), "6.28E-25");
   test(hasMoreTokens(scanner), false);
}

static void testScanEscapeSequences(void) {
   TokenScanner scanner;
   string token;

   trace(scanner = newTokenScanner());
   trace(scanStrings(scanner));
   trace(setInputString(scanner, "'A'"));
   trace(token = nextToken(scanner));
   test(token, "'A'");
   test(getStringValue(token), "A");
   trace(setInputString(scanner, "\"Quoted String\""));
   trace(token = nextToken(scanner));
   test(token, "\"Quoted String\"");
   test(getStringValue(token), "Quoted String");
   trace(setInputString(scanner, "\"\\\"Bother\\\" said Pooh.\""));
   trace(token = nextToken(scanner));
   test(token, "\"\\\"Bother\\\" said Pooh.\"");
   test(getStringValue(token), "\"Bother\" said Pooh.");
   trace(setInputString(scanner, "\"\\xFF\\007\\a\\t\""));
   trace(token = nextToken(scanner));
   test(token, "\"\\xFF\\007\\a\\t\"");
   test(getStringValue(token), "\377\007\a\t");
}

#endif

File: Archived/JavaBackEnd/latest-version-decompiled/acm/util/TokenScanner.java
package acm.util;

import java.io.*;
import java.util.ArrayList;
import java.util.Stack;

// Referenced classes of package acm.util:
//            ErrorException

public class TokenScanner
{

    public TokenScanner()
    {
        ignoreWhitespaceFlag = false;
        ignoreCommentsFlag = false;
        scanNumbersFlag = false;
        scanStringsFlag = false;
        cpos = 0;
        input = null;
        wordChars = "";
        savedTokens = new Stack();
        operators = new ArrayList();
    }

    public TokenScanner(String s)
    {
        this();
        setInput(s);
    }

    public TokenScanner(Reader reader)
    {
        this();
        setInput(reader);
    }

    public void setInput(String s)
    {
        setInput(((Reader) (new StringReader(s))));
    }

    public void setInput(Reader reader)
    {
        savedChars = "";
        savedTokens.clear();
        cpos = 0;
        input = reader;
    }

    public boolean hasMoreTokens()
    {
        String s = nextToken();
        saveToken(s);
        return !s.isEmpty();
    }

    public String nextToken()
    {
        if(!savedTokens.isEmpty())
            return (String)savedTokens.pop();
        else
            return scanToken();
    }

    public void saveToken(String s)
    {
        savedTokens.push(s);
    }

    public int getPosition()
    {
        if(savedTokens.isEmpty())
            return cpos;
        else
            return cpos - ((String)savedTokens.peek()).length();
    }

    public void ignoreWhitespace()
    {
        ignoreWhitespaceFlag = true;
    }

    public void ignoreComments()
    {
        ignoreCommentsFlag = true;
    }

    public void scanNumbers()
    {
        scanNumbersFlag = true;
    }

    public void scanStrings()
    {
        scanStringsFlag = true;
    }

    public void addWordCharacters(String s)
    {
        int i = 0;
_L3:
        if(i >= s.length()) goto _L2; else goto _L1
_L1:
        char c;
        c = s.charAt(i);
        if(wordChars.indexOf(c) != -1)
            continue; /* Loop/switch isn't completed */
        new StringBuilder();
        this;
        JVM INSTR dup_x1 ;
        wordChars;
        append();
        c;
        append();
        toString();
        wordChars;
        i++;
          goto _L3
_L2:
    }

    public boolean isWordCharacter(int i)
    {
        if(i == -1)
            return false;
        else
            return Character.isLetterOrDigit(i) || wordChars.indexOf((char)i) >= 0;
    }

    public void addOperator(String s)
    {
        operators.add(s);
    }

    public void verifyToken(String s)
    {
        String s1 = nextToken();
        if(s1.equals(s))
        {
            return;
        } else
        {
            String s2 = (new StringBuilder()).append("Found ").append(s1).append(" when expecting ").append(s).toString();
            throw new ErrorException(s2);
        }
    }

    public int getTokenType(String s)
    {
        if(s == null || s.length() == 0)
            return -1;
        char c = s.charAt(0);
        if(Character.isWhitespace(c))
            return 0;
        if(Character.isDigit(c))
            return 2;
        if(c == '"' || c == '\'')
            return 3;
        return !isWordCharacter(c) ? 4 : 1;
    }

    public int getChar()
    {
        char c;
        cpos++;
        if(savedChars.isEmpty())
            break MISSING_BLOCK_LABEL_43;
        c = savedChars.charAt(0);
        savedChars = savedChars.substring(1);
        return c;
        return input.read();
        IOException ioexception;
        ioexception;
        throw new ErrorException(ioexception);
    }

    public void ungetChar(int i)
    {
        cpos--;
        if(input instanceof PushbackReader)
            try
            {
                ((PushbackReader)input).unread(i);
                return;
            }
            catch(IOException ioexception) { }
        if(i >= 0)
            savedChars = (new StringBuilder()).append((char)i).append(savedChars).toString();
    }

    public String getStringValue(String s)
    {
        String s1 = "";
        boolean flag = false;
        int i = s.length();
        if(i > 1 && (s.charAt(0) == '"' || s.charAt(0) == '\''))
        {
            flag = true;
            i--;
        }
        for(int j = ((flag) ? 1 : 0); j < i; j++)
        {
            char c = s.charAt(j);
            if(c == '\\')
            {
                c = s.charAt(++j);
                if(Character.isDigit(c) || c == 'x')
                {
                    byte byte0 = 8;
                    byte byte1 = 3;
                    if(c == 'x')
                    {
                        byte0 = 16;
                        byte1 = 2;
                        j++;
                    }
                    int k = 0;
                    boolean flag1 = false;
                    for(int i1 = Math.min(i, j + byte1); j < i1; j++)
                    {
                        c = s.charAt(j);
                        int l;
                        if(Character.isDigit(c))
                            l = c - 48;
                        else
                        if(byte0 == 16 && c >= 'A' && c <= 'F')
                        {
                            l = (c - 65) + 10;
                        } else
                        {
                            if(byte0 != 16 || c < 'a' || c > 'f')
                                break;
                            l = (c - 97) + 10;
                        }
                        k = byte0 * k + l;
                    }

                    c = (char)k;
                    j--;
                } else
                {
                    switch(c)
                    {
                    case 97: // 'a'
                        c = '\007';
                        break;

                    case 98: // 'b'
                        c = '\b';
                        break;

                    case 102: // 'f'
                        c = '\f';
                        break;

                    case 110: // 'n'
                        c = '\n';
                        break;

                    case 114: // 'r'
                        c = '\r';
                        break;

                    case 116: // 't'
                        c = '\t';
                        break;

                    case 118: // 'v'
                        c = '\013';
                        break;

                    case 34: // '"'
                        c = '"';
                        break;

                    case 39: // '\''
                        c = '\'';
                        break;

                    case 92: // '\\'
                        c = '\\';
                        break;
                    }
                }
            }
            s1 = (new StringBuilder()).append(s1).append(c).toString();
        }

        return s1;
    }

    private String scanToken()
    {
        int i = scanChar();
        if(i == -1)
            return "";
        if(scanNumbersFlag && Character.isDigit(i))
        {
            ungetChar(i);
            return scanNumber();
        }
        if(isWordCharacter(i))
        {
            ungetChar(i);
            return scanWord();
        }
        if(scanStringsFlag && (i == 34 || i == 39))
        {
            ungetChar(i);
            return scanString();
        }
        String s = (new StringBuilder()).append("").append((char)i).toString();
        do
        {
            if(!isOperatorPrefix(s))
                break;
            int j = getChar();
            if(j == -1)
                break;
            s = (new StringBuilder()).append(s).append((char)j).toString();
        } while(true);
        for(; s.length() > 1 && !isOperator(s); s = s.substring(0, s.length() - 1))
            ungetChar(s.charAt(s.length() - 1));

        return s;
    }

    private int scanChar()
    {
        int i;
label0:
        {
            boolean flag = false;
            boolean flag1 = false;
            int k;
            do
            {
                do
                    i = getChar();
                while(ignoreWhitespaceFlag && Character.isWhitespace(i) && !isOperator(Character.toString((char)i)));
                if(flag)
                {
                    int j = getChar();
                    if(i == 42 && j == 47)
                        flag = false;
                    else
                        ungetChar(j);
                    continue;
                }
                if(flag1)
                {
                    if(i == 10 || i == 13)
                        flag1 = false;
                    continue;
                }
                if(!ignoreCommentsFlag || i != 47)
                    break label0;
                k = getChar();
                if(k == 42)
                {
                    flag = true;
                    continue;
                }
                if(k != 47)
                    break;
                flag1 = true;
            } while(true);
            ungetChar(k);
        }
        return i;
    }

    private String scanWord()
    {
        String s = "";
        int i;
        for(i = getChar(); isWordCharacter(i); i = getChar())
            s = (new StringBuilder()).append(s).append((char)i).toString();

        ungetChar(i);
        return s;
    }

    private String scanNumber()
    {
        String s = "";
        int i = 0;
        int j = 69;
        do
        {
            if(i == 6)
                break;
            int k = getChar();
            switch(i)
            {
            case 0: // '\0'
                if(k == 48)
                {
                    int l = getChar();
                    if(l == 120 || l == 88)
                    {
                        s = "0x";
                        continue;
                    }
                    ungetChar(l);
                }
                i = 1;
                break;

            case 1: // '\001'
                if(k == 46)
                    i = 2;
                else
                if(k == 69 || k == 101)
                    i = 3;
                else
                if(!Character.isDigit(k))
                    i = 6;
                break;

            case 2: // '\002'
                if(k == 69 || k == 101)
                {
                    j = k;
                    i = 3;
                } else
                if(!Character.isDigit(k))
                    i = 6;
                break;

            case 3: // '\003'
                if(k == 43 || k == 45)
                    i = 4;
                else
                if(Character.isDigit(k))
                    i = 5;
                else
                    i = 6;
                break;

            case 4: // '\004'
                if(Character.isDigit(k))
                {
                    i = 5;
                } else
                {
                    ungetChar(k);
                    k = j;
                    i = 6;
                }
                break;

            case 5: // '\005'
                if(!Character.isDigit(k))
                    i = 6;
                break;
            }
            if(i == 6)
            {
                ungetChar(k);
                break;
            }
            s = (new StringBuilder()).append(s).append((char)k).toString();
        } while(true);
        return s;
    }

    private String scanString()
    {
        char c = (char)getChar();
        String s = (new StringBuilder()).append("").append(c).toString();
        int i = 0;
        do
        {
            int j = getChar();
            if(j != -1 && (j != c || i == 92))
            {
                s = (new StringBuilder()).append(s).append((char)j).toString();
                i = j;
            } else
            {
                return (new StringBuilder()).append(s).append(c).toString();
            }
        } while(true);
    }

    private boolean isOperator(String s)
    {
        for(int i = 0; i < operators.size(); i++)
            if(((String)operators.get(i)).equals(s))
                return true;

        return false;
    }

    private boolean isOperatorPrefix(String s)
    {
        for(int i = 0; i < operators.size(); i++)
            if(((String)operators.get(i)).startsWith(s))
                return true;

        return false;
    }

    public static final int EOF = -1;
    public static final int SEPARATOR = 0;
    public static final int WORD = 1;
    public static final int NUMBER = 2;
    public static final int STRING = 3;
    public static final int OPERATOR = 4;
    private static final int INITIAL_STATE = 0;
    private static final int BEFORE_DECIMAL_POINT = 1;
    private static final int AFTER_DECIMAL_POINT = 2;
    private static final int STARTING_EXPONENT = 3;
    private static final int FOUND_EXPONENT_SIGN = 4;
    private static final int SCANNING_EXPONENT = 5;
    private static final int FINAL_STATE = 6;
    private Reader input;
    private String wordChars;
    private String savedChars;
    private Stack savedTokens;
    private ArrayList operators;
    private int cpos;
    private boolean ignoreWhitespaceFlag;
    private boolean ignoreCommentsFlag;
    private boolean scanNumbersFlag;
    private boolean scanStringsFlag;
}
